commons/zenoh-shm/src/api/provider/shared_memory_provider.rs

```diff
@@ -64,83 +67,69 @@ impl BusyChunk {
-    pub fn size(self, size: usize) -> AllocLayoutSizedBuilder<'a, IDSource, Backend> {
-        AllocLayoutSizedBuilder {
-            provider: self.provider,
-            size,
-        }
-    }
-}
 
 #[zenoh_macros::unstable_doc]
-pub struct AllocLayoutSizedBuilder<'a, IDSource, Backend>
+pub struct AllocLayoutSizedBuilder<'a, IDSource, Backend>(AllocData<'a, IDSource, Backend>)
 where
     IDSource: ProtocolIDSource,
-    Backend: SharedMemoryProviderBackend,
-{
-    provider: &'a SharedMemoryProvider<IDSource, Backend>,
-    size: usize,
-}
+    Backend: SharedMemoryProviderBackend;
+
 impl<'a, IDSource, Backend> AllocLayoutSizedBuilder<'a, IDSource, Backend>
 where
     IDSource: ProtocolIDSource,
     Backend: SharedMemoryProviderBackend,
 {
-    /// Set alignment for layout
+    fn new(provider: &'a SharedMemoryProvider<IDSource, Backend>, size: usize) -> Self {
+        Self(AllocData {
+            provider,
+            size,
+            alignment: AllocAlignment::default(),
+        })
+    }
+
+    /// Set alignment
     #[zenoh_macros::unstable_doc]
-    pub fn alignment(
-        self,
-        alignment: AllocAlignment,
-    ) -> AllocLayoutAlignedBuilder<'a, IDSource, Backend> {
-        AllocLayoutAlignedBuilder {
-            provider: self.provider,
-            size: self.size,
+    pub fn with_alignment(self, alignment: AllocAlignment) -> Self {
+        Self(AllocData {
+            provider: self.0.provider,
+            size: self.0.size,
             alignment,
-        }
+        })
     }
 
-    /// try to build an allocation layout
+    /// Try to build an allocation layout
     #[zenoh_macros::unstable_doc]
-    pub fn res(self) -> ZResult<AllocLayout<'a, IDSource, Backend>> {
-        AllocLayout::new(self.size, AllocAlignment::default(), self.provider)
+    pub fn make_layout(self) -> Result<AllocLayout<'a, IDSource, Backend>, ZLayoutError> {
+        AllocLayout::new(self.0)
     }
-}
 
-#[zenoh_macros::unstable_doc]
-pub struct AllocLayoutAlignedBuilder<'a, IDSource, Backend>
-where
-    IDSource: ProtocolIDSource,
-    Backend: SharedMemoryProviderBackend,
-{
-    provider: &'a SharedMemoryProvider<IDSource, Backend>,
-    size: usize,
-    alignment: AllocAlignment,
-}
-impl<'a, IDSource, Backend> AllocLayoutAlignedBuilder<'a, IDSource, Backend>
-where
-    IDSource: ProtocolIDSource,
-    Backend: SharedMemoryProviderBackend,
-{
-    /// Try to build layout with specified args
+    /// Set the allocation policy
     #[zenoh_macros::unstable_doc]
-    pub fn res(self) -> ZResult<AllocLayout<'a, IDSource, Backend>> {
-        AllocLayout::new(self.size, self.alignment, self.provider)
+    pub fn with_policy<Policy>(self) -> AllocBuilder2<'a, IDSource, Backend, Policy> {
+        AllocBuilder2 {
+            data: self.0,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Get the result
+    #[zenoh_macros::unstable_doc]
+    pub fn res(self) -> BufLayoutAllocResult {
```

Comment:
https://github.com/eclipse-zenoh/zenoh/pull/942 has been recently merged. I believe we should use either `.wait()` or `.await` from now on.

Reply:
I know about our new `Wait` and `IntoFuture` traits, I was just concerned that `.wait()` will confuse our users that it blocks when it doesn't.

Practically it blocks only if BlockOn policy is used, so we have three types of behaviors:
- non-blocking (any policy except BlockOn)
- blocking (only BlockOn policy):
- - sync
- - async (available for `Backend: Sync`)

I think I need to split allocation API into these three forms:
- `res()` for non-blocking policies
- `Wait` for sync BlockOn
- `IntoFuture` for async BlockOn

Reply:
Regardless it's blocking or non-blocking, `.wait()` is the common method for resolving builders in `sync` context in the same way you call `.await` in `async` context. I don't see much benefit for adding another way to resolve the builder...